### ISSUE TITLE
spike: propose solution for installed packages not showing up in path

### ISSUE THREAD

**Issue by mkenigs** [flox/flox#1030]
Bug copied from https://github.com/flox/flox/issues/810

STR:
be on a macOS machine

* have nvm on your system and a node version installed 
* `flox init`
* `flox activate`
* `flox install nodejs`
* `which node`

Expected: flox's node
Actual: nvm's node

Propose an approach to solve this

**Comment by aakropotkin** [comment]
can you provide more detail about the existing `PATH` setting?

**Comment by mkenigs** [comment]
I think it's because of `hash`, not `PATH`
```
flox [tmp] /tmp  > which ls
/bin/ls
flox [tmp] /tmp  > flox install coreutils
✅ 'coreutils' installed to environment tmp at /private/tmp
flox [tmp] /tmp  > which ls
/bin/ls
flox [tmp] /tmp  > echo $PATH
/private/tmp/.flox/run/aarch64-darwin.tmp/bin:...
flox [tmp] /tmp  > hash -r
flox [tmp] /tmp  > which ls
/private/tmp/.flox/run/aarch64-darwin.tmp/bin/ls
```

**Comment by mkenigs** [comment]
I'm not seeing this on Linux+bash, Linux+zsh, or macOS+bash. So I think all we need is on macOS+zsh, to add something like this to our zsh activation:
```
_flox_hook() {
  trap -- '' SIGINT;
  # if we haven't done it already, for every path in "$FLOX_ENV/bin", hash basename=path
  trap - SIGINT;
}
typeset -ag precmd_functions;
if [[ -z "${precmd_functions[(r)_flox_hook]+1}" ]]; then
  precmd_functions=( _flox_hook ${precmd_functions[@]} )
fi
typeset -ag chpwd_functions;
if [[ -z "${chpwd_functions[(r)_flox_hook]+1}" ]]; then
  chpwd_functions=( _flox_hook ${chpwd_functions[@]} )
fi
```

**Comment by ysndr** [comment]
Works on macos for me, though I am using a nix provided zsh...

**Comment by tomberek** [comment]
https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/programs/bash/bash.nix#L135

**Comment by mkenigs** [comment]
> https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/programs/bash/bash.nix#L135

That's pretty invasive

**Comment by aakropotkin** [comment]
You know about the wonky path fixup stuff for `/bin/zsh` on OSX right?
Assuming you knew that already, have you tried _the wrapper script_ trick that people usually use?

Essentially `if` [[ "$( uname -s )" = 'Darwin' ]] && [[ "$SHELL" = '/bin/zsh' ]] then $SHELL -i -c 'export PATH="$FLOX_ENV/bin:$PATH"; . "$FLOX_ENV/activate.sh"; else ...; fi`

**Comment by mkenigs** [comment]
Yeah I think that's a separate issue https://github.com/flox/flox/issues/811. Like I showed above, in this case we're not getting a binary from the front of `PATH`

**Comment by mkenigs** [comment]
We should be able to fix this with:

Add something like this to our zsh activation:
```
real_flox="$(whence flox)"
source <(cat << EOF
flox() {
  "$real_flox" "\$@"
  export PATH="$PATH"
}
EOF
)
```

Do something similar for bash
